pkg/flow/hcltypes/secret.go

```diff
@@ -0,0 +1,75 @@
+package hcltypes
+
+import (
+	"reflect"
+
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/rfratto/gohcl"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// Secret holds a sensitive value. Secrets are never displayed to the user when
+// rendering HCL.
+//
+// String values may be implicitly converted to a Secret, but the inverse isn't
+// true: this ensures that it's not possible for a user to accidentally leak a
+// sensitive value.
```

Comment:
I'm not sure what we mean by this. We certainly can do the following

```go
s := Secret("hello world!")
fmt.Println(string(s)) // hello world!
```

This is tied to the general question about how the users/developers will interact with these types. People who write components will make use of Secret/OptionalSecret and the unexported secretTy/optionalSecretTy are to be used when encoding/decoding HCL, right?

Reply:
On second reading, I guess it means that it can't be converted back when referred to as arguments and exports of a Flow component which are go-cty values in the background.

Reply:
> People who write components will make use of Secret/OptionalSecret and the unexported secretTy/optionalSecretTy are to be used when encoding/decoding HCL, right?

Right, component authors will use Secret/OptionalSecret directly and that will be converted to registered cty capsule types under the hood. 

> But will they also handle conversions between these types manually? Or is this also something that's going to happen on the HCL layer as Flow components pass values between each other?

Conversions happen automatically when decoding. A component author could expect a password to be a Secret and they don't have to know whether the final value came from a Secret or a string. 

> On second reading, I guess it means that it can't be converted back when referred to as arguments and exports of a Flow component which are go-cty values in the background.

Yeah, I'm talking about the go-cty conversion being disallowed so users can't convert (but developers can). I'll update the comment to try to make that more clear.

Reply:
Thanks, made more sense after a second reading! I played around with the tests a little, and I don't have anything to add ^^